activerecord/lib/arel/visitors/mysql.rb

```diff
@@ -62,7 +62,7 @@ def visit_Arel_Nodes_NotRegexp(o, collector)
         # query. However, this does not allow for LIMIT, OFFSET and ORDER. To support
         # these, we must use a subquery.
         def prepare_update_statement(o)
-          if o.offset || has_join_sources?(o) && has_limit_or_offset_or_orders?(o)
+          if o.offset || (has_join_sources?(o) && has_limit_or_offset_or_orders?(o)) || has_join_sources?(o)
```

Comment:
This change has a regression for 859fba7c4bf7d33b4f9655914ed4bdc85380552e.
I've pushed a regression test for that to master b535a455a8a8e83901c8720eac9b0cb4a6445680.

Reply:
@kamipo So, do i need to make any changes in this pull request? 
Could you please provide guidance.

Reply:
Unlike `ORDER`, `LIMIT`, and `JOIN`, AFAICS `GROUP BY` and `HAVING` are not valid clauses for UPDATE and DELETE statements in all supported RDBMSes (the formers are valid clauses for UPDATE and DELETE statements in MySQL, that is the reason why those are had in the statements to avoid subselect generation).

So I'd suggest to push conditions as subselect from the begining if a relation has a `GROUP BY` or `HAVING`.